OMCompiler/SimulationRuntime/c/simulation/solver/gbode_util.c

```diff
@@ -0,0 +1,544 @@
+/*
+ * This file is part of OpenModelica.
+ *
+ * Copyright (c) 1998-2022, Open Source Modelica Consortium (OSMC),
+ * c/o Linköpings universitet, Department of Computer and Information Science,
+ * SE-58183 Linköping, Sweden.
+ *
+ * All rights reserved.
+ *
+ * THIS PROGRAM IS PROVIDED UNDER THE TERMS OF THE BSD NEW LICENSE OR THE
+ * GPL VERSION 3 LICENSE OR THE OSMC PUBLIC LICENSE (OSMC-PL) VERSION 1.2.
+ * ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS PROGRAM CONSTITUTES
+ * RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
+ * ACCORDING TO RECIPIENTS CHOICE.
+ *
+ * The OpenModelica software and the OSMC (Open Source Modelica Consortium)
+ * Public License (OSMC-PL) are obtained from OSMC, either from the above
+ * address, from the URLs: http://www.openmodelica.org or
+ * http://www.ida.liu.se/projects/OpenModelica, and in the OpenModelica
+ * distribution. GNU version 3 is obtained from:
+ * http://www.gnu.org/copyleft/gpl.html. The New BSD License is obtained from:
+ * http://www.opensource.org/licenses/BSD-3-Clause.
+ *
+ * This program is distributed WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, EXCEPT AS
+ * EXPRESSLY SET FORTH IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE
+ * CONDITIONS OF OSMC-PL.
+ *
+ */
+
+/*! \file gbode_util.c
+ */
+#include "gbode_util.h"
+
+
+// LA functions
+void addSmultVec_gbf(double* a, double* b, double *c, double s, int nIdx, int* idx) {
+  int i, ii;
+
+  for (ii=0; ii<nIdx; ii++) {
+    i = idx[ii];
+    a[i] = b[i] + s*c[i];
+  }
+}
+
+void addSmultVec_gb(double* a, double* b, double *c, double s, int n) {
+  int i;
+
+  for (i=0; i<n; i++) {
+    a[i] = b[i] + s*c[i];
+  }
+}
+
+// interpolation function
+
+/**
+ * @brief Linear interpolation of all vector components
+ *
+ * @param ta      Time value at the left hand side
+ * @param fa      Function values at the left hand side
+ * @param tb      Time value at the right hand side
+ * @param fb      Function values at the right hand side
+ * @param t       Time value at the interpolated time point
+ * @param f       Function values at the interpolated time point
+ * @param n       Size of the vector
+ */
+void linear_interpolation_gb(double ta, double* fa, double tb, double* fb, double t, double* f, int n)
```

Comment:
Unify `linear_interpolation_gb` and `linear_interpolation_gbf` as well as `error_interpolation_gb` and `error_interpolation_gbf`.